dpe/src/dpe_instance.rs

```diff
@@ -389,6 +389,83 @@ impl DpeInstance {
 
         hasher.finish().map_err(|_| DpeErrorCode::HashError)
     }
+
+    /// Retrieve index of the root context based on the tci_type and locality of
+    /// the context matching the inputs
+    ///
+    /// # Arguments
+    ///
+    /// * `idx` - desired index of the root context
+    /// * `locality` - desired locality of the root context
+    pub fn get_root_idx(&self, tci_type: u32, locality: u32) -> Result<usize, DpeErrorCode> {
+        // find indices of all non-inactive contexts matching tci_type and locality
+        let mut tci_nodes = self.contexts.iter().enumerate().filter(|(_, context)| {
+            context.state != ContextState::Inactive
+                && context.locality == locality
+                && context.tci.tci_type == tci_type
+        });
+        let tci = tci_nodes.next();
+        // ensure there is exactly 1 match
+        if tci.is_none() || tci_nodes.next().is_some() {
+            return Err(DpeErrorCode::InternalError);
+        }
+        Ok(tci.unwrap().0)
+    }
+
+    /// Retrieve tci_current of a context
+    ///
+    /// # Arguments
+    ///
+    /// * `idx` - index of the context
+    pub fn get_current_tci(&self, idx: usize) -> Result<TciMeasurement, DpeErrorCode> {
+        if idx >= MAX_HANDLES {
+            return Err(DpeErrorCode::InternalError);
+        }
+        Ok(self.contexts[idx].tci.tci_current)
+    }
+
+    /// Determines if the context array represents a valid tree by checking that
+    /// there is only 1 connected component and that all nodes lead up to
+    /// the root node.
+    ///
+    /// # Arguments
+    ///
+    /// * `root_idx` - The index of the root context
+    pub fn validate_context_tree(&self, root_idx: usize) -> bool {
```

Comment:
Ideally we would test this on a large number of inputs for how the Context array is populated. But this is a hard function to fuzz because the test has no way to check whether the answer this function gives is the correct answer for the generated output.

I don't have a solution, and we don't have to solve in this PR, just something to think about for how we could improve testing here.

Reply:
Thanks for mentioning this, I made https://github.com/chipsalliance/caliptra-dpe/issues/204 to track.